contracts/ERC2470.sol

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: CC0-1.0
+pragma solidity 0.7.3;
+
+
+/**
+ * @title ERC-2470: Singleton Factory
+ * @notice Exposes CREATE2 (EIP-1014) to deploy bytecode on deterministic addresses based on initialization code and salt.
+ * @author Ricardo Guilherme Schmidt <ricardo3@status.im>
+ * @author Richard Meissner <richard@gnosis.io>
+ */
+contract ERC2470 {
+
+    /**
+     * @notice Fails on calls with no `msg.data` but with `msg.value`
+     */
+    receive() external payable {
+        revert("ERC2470: bad call");
+    }
+
+    /**
+     * @notice Fallback forwards `msg.data` with salt `bytes32(0)` to `deploy(bytes,bytes32)`.
+     */
+    fallback() external payable {
+        address payable createdContract = deploy(msg.data, bytes32(0));
+        assembly {
+            mstore(0, createdContract)
+            return(0, 32)
+        }
+    }
+
+    /**
+     * @notice Deploys `_initCode` using `_salt` for defining the deterministic address.
+     * @param _initCode Initialization code.
+     * @param _salt Arbitrary value to modify resulting address.
+     * @return createdContract Created contract address.
+     */
+    function deploy(bytes memory _initCode, bytes32 _salt)
+        public
+        payable
+        returns (address payable createdContract)
+    {
+        assembly {
+            createdContract := create2(callvalue(), add(_initCode, 0x20), mload(_initCode), _salt)
+        }
+        require(createdContract != address(0), "ERC2470: CREATE2 failed");
+    }
+
+    /**
+     * @notice Calculates the resulting address of a contract created in this factory.
+     * @param _initCode Initialization code.
+     * @param _salt Arbitrary value to modify resulting address.
+     * @return resultAddress predicted address.
+     */
+    function predictedAddress(
```

Comment:
Just out of curiousity, how much does adding this method increase the costs to deploy a contract?

Reply:
With predictAddress: 269084 gas.
Without:  108553 gas
Delta: **160531 gas**.

Thats why the first version didn't had this as is very easy to calculate this on frontend.

Reply:
The only reason for having this inside of this contract is for acting like a library to other contracts, so they don't have to include this function on their code. 

However, a more generic library could be used by these contracts, which predictedAddress of any create2 contract.